python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py

```diff
@@ -2120,74 +2192,82 @@ def set_concurrency_slots(self, concurrency_key: str, num: int) -> None:
         if num < 0:
             raise DagsterInvalidInvocationError("Cannot have a negative number of slots.")
 
-        keys_to_assign = None
+        # ensure that we have concurrency limits set for all keys
+        self._reconcile_concurrency_limits_from_slots()
+
         with self.index_connection() as conn:
-            count_row = conn.execute(
-                db_select([db.func.count()])
+            self._upsert_and_lock_limit_row(conn, concurrency_key, num)
+            keys_to_assign = self._allocate_concurrency_slots(conn, concurrency_key, num)
+
+        if keys_to_assign:
+            # we've added some slots... if there are any pending steps, we can assign them now or
+            # they will be unutilized until free_concurrency_slots is called
+            self.assign_pending_steps(keys_to_assign)
+
+    def _allocate_concurrency_slots(self, conn, concurrency_key: str, num: int) -> List[str]:
+        keys_to_assign = []
+        count_row = conn.execute(
+            db_select([db.func.count()])
+            .select_from(ConcurrencySlotsTable)
+            .where(
+                db.and_(
+                    ConcurrencySlotsTable.c.concurrency_key == concurrency_key,
+                    ConcurrencySlotsTable.c.deleted == False,  # noqa: E712
+                )
+            )
+        ).fetchone()
+        existing = cast(int, count_row[0]) if count_row else 0
+
+        if existing > num:
+            # need to delete some slots, favoring ones where the slot is unallocated
```

Comment:
I don't think so... we mark them as "deleted" so that they won't be allocated anymore, but retain the record until the step that has occupied it has been freed.